exercise/map.js

```diff
@@ -0,0 +1,56 @@
+/*
+## Map
+
+### Instructions
+
+Create an array named 'arr2' that contains the values of 'arr1'
+with the following modifications applied to each of the values:
+1) They have to be multiplied by 2
+2) Then the results have to be subtracted by 4
+3) finally, the negative values have to be replaced by null
+then do the same with arrRandom1 by returning arrRandom2
+The use of `for` and `while` is forbidden for this exercise
+
+### Notions
+
+- Map
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
+
+*/
+let arr1 = [10, -10, 20, -95, 86]
+///*/ // ⚡
+
+///*/// ⚡
+
+let arrRandom1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+
+for (let i = 0; i < 10; i++) {
+  let value = Math.round(Math.random() * (10 - -10 - 10))
+  arrRandom1.push(value)
+}
+
+export const tests = []
+const t = (f) => tests.push(f)
+
+const check = (arg) => arg * 2 - 4
+
+t(() => arr2[0] === check(arr1[0]))
+t(() => arr2[1] === null)
+t(() => arr2[2] === check(arr1[2]))
+t(() => arr2[3] === null)
+t(() => arr2[4] === check(arr1[4]))
+
+t(() => {
+  for (let i of arrRandom1) {
+    if (arrRandom1[i] >= 2)
+      if (arrRandom2[i] !== check(arrRandom1[i])) return false
+    if (arrRandom1[i] < 2) if (arrRandom2[i] !== null) return false
+    return true
+  }
+})
+
+t(({ code }) => !code.includes('for'))
+t(({ code }) => !code.includes('while'))
```

Comment:
I think a better way to ensure they used map would be to "monkey patch" map:

```js
t((_, ctx) => {
  ctx.mapCall = []
  const { map } = Array.prototype
  Array.prototype.map = function () {
    ctx.mapCalls.push([this, ...arguments])
    return map.apply(this, arguments)
  }
})
```
now this wierd shitty code allow to wrap the *classic* `[].map` in a function that save the arguments of each calls.

so in your tests you can check in the context if they used map or not, and try to mimic the call:
```js
t(({ eq }, ctx) => eq(fn([10, -2, 4]), [16, null, 0]) // 1st test execute it
t(({ eq }, ctx) => eq(ctx.mapCalls[ctx.mapCalls.length - 1][0], [10, -2, 4]) // check that map was called on this array
```

Reply:
In use now. I transmitted the knowledge to Afonso.